# First Things First

To start with, I'm simply going to build a basic web crawler then, using the domains from the URDB source URLs file, I'll randomly sample 1000 to see what their robots.txt file says (to keep it simple, I'll parse them first to see which ones are empty/missing, then check to see what the non-missing/non-empty ones say. For reference, the standards for constructing a robots.txt file [is explained here](http://www.robotstxt.org/robotstxt.html).

In [1]:
import scrapy
import pandas as pd

In [2]:
URLs = pd.read_table('URDB_URLs_9-24-18.txt', names = ['Tariff Sheet URL'])
URLs

Tariff Sheet URL
0      http://www.iid.com/Modules/ShowDocument.aspx?d...
1      http://villageofarcade.org/departments/public-...
2      http://cecelec.coopwebbuilder.com/sites/cecele...
3      http://www.wakeforestnc.gov/client_resources/r...
4         http://www.rrvcoop.com/content/security-lights
5      http://www.prairielandelectric.com/Rates_PDF/M...
6         http://www.talquinelectric.com/rates_elec.aspx
7      http://www.midstateelectric.coop/customer-serv...
8      http://www.surprisevalleyelectric.org/_uploads...
9      http://www.mainepublicservice.com/media/40402/...
10     http://www.ecirec.coop/admin/upload/editor/fil...
11     http://psc.wi.gov/apps40/tariffs/viewfile.aspx...
12          http://www.singingriver.com/Files/LGS-15.pdf
13     http://www.delaware.coop/sites/default/files/T...
14     http://www.iid.com/Modules/ShowDocument.aspx?d...
15        http://www.bartlettec.coop/ServiceandFees.aspx
16     http://www.clatskaniepud.com/index.php/rate-sc...
17     http://www.sanpatricioelectric.org/electricser...
18                       http://www.npec.org/slrates.htm
19     http://www.neelectric.com/pdf/rate-large-comme...
20     http://sfwater.org/modules/showdocument.aspx?d...
21        http://www.brmemc.com/Electric_Rate_Sheet.aspx
22     http://www.cme.coop/Member_Services/rates/Sche...
23     https://www.swepco.com/global/utilities/lib/do...
24        http://www.brmemc.com/Electric_Rate_Sheet.aspx
25     http://www.iplpower.com/Our_Company/Rates/Rate...
26                                 http://www.iclp.coop/
27     http://www.claycountyelectric.com/content/summ...
28     http://lodielectric.com/commercial/rateform.as...
29     http://www.morristownutilities.org/General_Pow...
...                                                  ...
39040  https://www.nyseg.com/wps/wcm/connect/16c28c6c...
39041  https://www.nyseg.com/wps/wcm/connect/16c28c6c...
39042  https://www.lipower.org/wp-content/uploads/201...
39043  https://www.lipower.org/wp-content/uploads/201...
39044  https://www.lipower.org/wp-content/uploads/201...
39045  https://www.lipower.org/wp-content/uploads/201...
39046  https://www.lipower.org/wp-content/uploads/201...
39047  https://www.lipower.org/wp-content/uploads/201...
39048  https://www.lipower.org/wp-content/uploads/201...
39049  https://www.lipower.org/wp-content/uploads/201...
39050  https://www.lipower.org/wp-content/uploads/201...
39051  https://www.lipower.org/wp-content/uploads/201...
39052  https://www.lipower.org/wp-content/uploads/201...
39053  https://www.lipower.org/wp-content/uploads/201...
39054  https://www.lipower.org/wp-content/uploads/201...
39055  https://www.lipower.org/wp-content/uploads/201...
39056  https://www.lipower.org/wp-content/uploads/201...
39057  https://www.lipower.org/wp-content/uploads/201...
39058  https://www.lipower.org/wp-content/uploads/201...
39059  https://www.lipower.org/wp-content/uploads/201...
39060  https://www.lipower.org/wp-content/uploads/201...
39061  https://www.lipower.org/wp-content/uploads/201...
39062  https://www.lipower.org/wp-content/uploads/201...
39063  https://www.lipower.org/wp-content/uploads/201...
39064  https://www.lipower.org/wp-content/uploads/201...
39065  https://www.lipower.org/wp-content/uploads/201...
39066  https://www.lipower.org/wp-content/uploads/201...
39067  https://www.lipower.org/wp-content/uploads/201...
39068  https://www.lipower.org/wp-content/uploads/201...
39069  https://www.lipower.org/wp-content/uploads/201...

[39070 rows x 1 columns]

# The Procedure

Based upon some very brief looking-over of these data, it seems like I'll need to do a few things to clean these URLs up first:

1. Check how many unique values there are relative to the DataFrame size
2. Reduce all URLs to their primary domain (e.g. https://www.nyseg.com/)
3. Remove all duplicates
4. Slap "robots.txt" on the end
5. Drop duplicate URLs again, since we probably now have more!
5. Run these through scrapy

In [3]:
URLs.drop_duplicates(inplace = True)

#Find only the domain portion of the URL + http... and then add /robots.txt to the end
URLs['Utility Domain URL'] = URLs['Tariff Sheet URL'].str.extract(pat = r'(https*://[^/]+)')
URLs['robots.txt URL'] = URLs['Utility Domain URL'] + '/robots.txt'

URLs.drop_duplicates(inplace = True)
URLs

Tariff Sheet URL  \
0      http://www.iid.com/Modules/ShowDocument.aspx?d...   
1      http://villageofarcade.org/departments/public-...   
2      http://cecelec.coopwebbuilder.com/sites/cecele...   
3      http://www.wakeforestnc.gov/client_resources/r...   
4         http://www.rrvcoop.com/content/security-lights   
5      http://www.prairielandelectric.com/Rates_PDF/M...   
6         http://www.talquinelectric.com/rates_elec.aspx   
7      http://www.midstateelectric.coop/customer-serv...   
8      http://www.surprisevalleyelectric.org/_uploads...   
9      http://www.mainepublicservice.com/media/40402/...   
10     http://www.ecirec.coop/admin/upload/editor/fil...   
11     http://psc.wi.gov/apps40/tariffs/viewfile.aspx...   
12          http://www.singingriver.com/Files/LGS-15.pdf   
13     http://www.delaware.coop/sites/default/files/T...   
14     http://www.iid.com/Modules/ShowDocument.aspx?d...   
15        http://www.bartlettec.coop/ServiceandFees.aspx   
16     http://www.clatskaniepud.com/index.php/rate-sc...   
17     http://www.sanpatricioelectric.org/electricser...   
18                       http://www.npec.org/slrates.htm   
19     http://www.neelectric.com/pdf/rate-large-comme...   
20     http://sfwater.org/modules/showdocument.aspx?d...   
21        http://www.brmemc.com/Electric_Rate_Sheet.aspx   
22     http://www.cme.coop/Member_Services/rates/Sche...   
23     https://www.swepco.com/global/utilities/lib/do...   
25     http://www.iplpower.com/Our_Company/Rates/Rate...   
26                                 http://www.iclp.coop/   
27     http://www.claycountyelectric.com/content/summ...   
28     http://lodielectric.com/commercial/rateform.as...   
29     http://www.morristownutilities.org/General_Pow...   
30     https://www.swepco.com/global/utilities/lib/do...   
...                                                  ...   
38433  https://oge.com/wps/wcm/connect/1354757f-c19f-...   
38434  http://www.oge.com/Documents/ARK/2011%20Arkans...   
38438  https://oge.com/wps/wcm/connect/de176c54-2711-...   
38439  http://oge.com/Documents/ARK/2011%20Arkansas%2...   
38440  http://www.oge.com/Documents/OK/6.40%20GS-TOU.pdf   
38441  https://oge.com/wps/wcm/connect/ef1b9f4d-6213-...   
38452  https://www.chelanpud.org/docs/default-source/...   
38462  https://www.eversource.com/content/docs/defaul...   
38514        https://www.otpco.com/media/1384/nd_901.pdf   
38515        https://www.otpco.com/media/1385/nd_902.pdf   
38516       https://www.otpco.com/media/1387/nd_1001.pdf   
38519       https://www.otpco.com/media/1388/nd_1002.pdf   
38521       https://www.otpco.com/media/1389/nd_1003.pdf   
38524       https://www.otpco.com/media/1392/nd_1005.pdf   
38528       https://www.otpco.com/media/1353/nd_1102.pdf   
38530       https://www.otpco.com/media/1356/nd_1105.pdf   
38541  https://www.blackhillsenergy.com/sites/default...   
38553    http://www.blackhillspower.com/pdf/SDTariff.pdf   
38624  https://www.consumersenergy.com/-/media/CE/Doc...   
38677  http://www.michigan.gov/documents/mpsc/dtee1cu...   
38794  https://www.aepohio.com/global/utilities/lib/d...   
38795  https://www2.dps.ny.gov/ETS/jobs/display/downl...   
38820  https://www.aepohio.com/global/utilities/lib/d...   
38822  https://nj.pseg.com/-/media/pseg/public-site/d...   
38836  https://www.firstenergycorp.com/content/dam/cu...   
38858  https://www.firstenergycorp.com/content/dam/cu...   
38945  https://www.peco.com/SiteCollectionDocuments/R...   
38949  https://www.peco.com/SiteCollectionDocuments/C...   
38994  https://www.nyseg.com/wps/wcm/connect/16c28c6c...   
39042  https://www.lipower.org/wp-content/uploads/201...   

                          Utility Domain URL  \
0                         http://www.iid.com   
1                 http://villageofarcade.org   
2          http://cecelec.coopwebbuilder.com   
3                http://www.wakeforestnc.gov   
4                     http://www.rrvcoop.com   
5         http://www.prairielan

In [4]:
#URLs['Tariff Sheet URL'].str.split(pat = r'^https*://.+/', expand = True)
#This doesn't do what I want right now, but gets the filenames, so maybe useful down the road?

In [5]:
URLs['robots.txt URL'].nunique()

1968

In [6]:
URLs.isnull().sum()

Tariff Sheet URL        0
Utility Domain URL    228
robots.txt URL        228
dtype: int64

In [7]:
URLs['Crawling_Allowed'] = False

In [9]:
URLs['Num_Ratepayers'] = 0

# TO DO
1. Figure out why these 228 nulls exist and correct them
2. Figure out rules for what we care about in robots.txt file
    1. Check if file exists, if not set `Crawling_Allowed = True`
    2. Check if file is blank. If so, set `Crawling_Allowed = True`
    2. If file doesn't have a line `User-agent: *`, then set `Crawling_Allowed = True`. This is fine because our bot doesn't exist yet and thus couldn't be disallowed or allowed by name in the first place.
    3. If the code below is in robots.txt and nothing else, set `Crawling_Allowed = True`
        `User-agent: *`
        `Disallow: `
    4. Check for the code below in file. If it exists, move on.
        `User-agent: *`
        `Disallow: /`
    5. For those remaining with `User-agent: *`:
        1. Compare our tariff sheet URL one-level-up parent directories to what is disallowed and set `Crawling_Allowed = True` if our directory isn't disallowed explicitly
3. For those with `Crawling_Allowed = True`, use their `Num_Ratepayers` value their score and final metric of "% Allowed" will be `(URLs['Num_Ratepayers'] * URLs['Crawling_Allowed']).sum()/URLs['Num_Ratepayers'].sum()`